src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -2923,10 +2932,109 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
+            }else{
+                throw SemanticError("Only Name supported as target in AnnAssign inside struct", x.m_body[i]->base.loc);
+            }
+            ASR::expr_t* init_expr = nullptr;
+            ASR::abiType abi=ASR::abiType::Source;
+            bool is_allocatable = false, is_const = false;
+            ASR::ttype_t *type = nullptr;
+            type = ast_expr_to_asr_type(ann_assign.m_annotation->base.loc, *ann_assign.m_annotation, is_allocatable, is_const, true);
+            if (ASR::is_a<ASR::FunctionType_t>(*type)) {
+                ASR::FunctionType_t* fn_type = ASR::down_cast<ASR::FunctionType_t>(type);
+                handle_lambda_function_declaration(var_name, fn_type, ann_assign.m_value, ann_assign.base.base.loc);
+                return;
+            }
+            ASR::storage_typeType storage_type = ASR::storage_typeType::Default;
+            if (is_allocatable) {
+                type = ASRUtils::TYPE(ASR::make_Allocatable_t(al, type->base.loc,
+                    ASRUtils::type_get_past_pointer(type)));
+            }
+            if (is_const) {
+                storage_type = ASR::storage_typeType::Parameter;
+            }
+
+            create_add_variable_to_scope(var_name, type,
+                    ann_assign.base.base.loc, abi, storage_type);
+
+            tmp = nullptr;
+            if (ann_assign.m_value) {
+                this->visit_expr(*ann_assign.m_value);
+            } else {
+                if (ASR::is_a<ASR::StructType_t>(*type)) {
+                    //`s` must be initialized with an instance of S
+                    throw  SemanticError("`" + var_name + "` must be initialized with an instance of " +
+                            ASRUtils::type_to_str_python(type), ann_assign.base.base.loc);
+                }
+            }
+            if (tmp && ASR::is_a<ASR::expr_t>(*tmp)) {
+                ASR::expr_t* value = ASRUtils::EXPR(tmp);
+                ASR::ttype_t* underlying_type = type;
+                cast_helper(underlying_type, value, value->base.loc);
+                if (!ASRUtils::check_equal_type(underlying_type, ASRUtils::expr_type(value), true)) {
+                    std::string ltype = ASRUtils::type_to_str_python(underlying_type);
+                    std::string rtype = ASRUtils::type_to_str_python(ASRUtils::expr_type(value));
+                    diag.add(diag::Diagnostic(
+                        "Type mismatch in annotation-assignment, the types must be compatible",
+                        diag::Level::Error, diag::Stage::Semantic, {
+                            diag::Label("type mismatch ('" + ltype + "' and '" + rtype + "')",
+                                    {ann_assign.m_target->base.loc, value->base.loc})
+                        })
+                    );
+                    throw SemanticAbort();
+                }
+                init_expr = value;
+            }
+            
+            if ( is_const ) {
```

Comment:
Same, remove all the lines that are not tested. Like, const, allocatable, cptr, ...